js/runner.go

```diff
@@ -765,6 +766,16 @@ func (u *ActiveVU) RunOnce() error {
 	ctx, cancel := context.WithCancel(u.RunContext)
 	defer cancel()
 	u.moduleVUImpl.ctx = ctx
+
+	eventIterData := event.IterData{
+		Iteration:    uint64(u.iteration),
+		VUID:         u.ID,
+		ScenarioName: u.scenarioName,
+	}
+	if u.Runner.preInitState.Events != nil {
+		u.Runner.preInitState.Events.Emit(&event.Event{Type: event.IterStart, Data: eventIterData})
```

Comment:
@ka3de Is the plan still to initialize browsers on the `IterStart` event?

We currently don't block to wait for the event to be processed here before running the iteration, since that could add a delay not just to the first iteration, but all.

Would it be possible for you to initialize browsers on the `Init` event, on which we do wait for? It's meant for that purpose, but I'm not sure if this aligns with your current approach.

Reply:
Hey @imiric , apologies for the delay, we (me or someone else from the team) still have pending to review this PR properly.

> Is the plan still to initialize browsers on the IterStart event?

That is still the plan, as we are currently using this _"1 browser process per iteration"_ model. I think initializing everything on the `Init` event would not be ideal, as that, with the current model, would mean to start up _N_ browsers initially, when most of these will be "unused" until the corresponding iteration for which it was allocated starts.

Nevertheless I understand non blocking on the `IterStart` event processing and I think we would be able to work around this from our side anyway.

Any thoughts on this @ankur22 , @inancgumus ?

Reply:
If i remember correctly, there's no way to retrieve the `scenario.options.browser` options when any of these events are fired, since the only way currently to retrieve these options is from the `vu` when an API is called. If that's the case then @ka3de has outlined everything already 👍

Reply:
> there's no way to retrieve the `scenario.options.browser` options when any of these events are fired

Ok, that's something that I overlooked, because that means that in the future, when we support more than one browser type, we won't be able to start the browser on `IterStart` event as we don't have access to that setting at that time.
In any case, by now this is an improvement and the biggest dependency was with `IterEnd`/`Exit` as this will allow us to do the cleanup.

Reply:
Alright, sounds good.

Ankur is right, but you _can_ access `vu.State.Options` on `IterStart` and `IterEnd`, since the `vu.State` is accessible at that point. Technically, the options are available to k6 when we emit `Init` as well, and a hack could be sending them as part of the `Init` event `Data`. This might not be such a hack after all, if we consider all `Init` subscribers to require options for initialization. :thinking:

This would literally be a one-line change to do, and would avoid any complicated refactoring to pass options in a dedicated API. We don't need to do this now, just pointing out that it's a possibility. FYI @mstoykov

Reply:
I added waiting for `IterStart` and `IterEnd` to be processed in a97bdd057, but as mentioned [here](https://github.com/grafana/k6/pull/3112#discussion_r1237004171), this has an impact on performance.

Let's continue the discussion in that thread.